### PR TITLE
xds-failover: add stat tracking primary/failover connection

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: xds-failover: add stat tracking primary/failover connection
Additional Description:
Extends the possible values of the `control_plane.connected_state` gauge. Previously it only stored 0 when Envoy wasn't connecting/connected to the xDS server, and 1 when it was.
This PR changes it to have the following values:
0 - Envoy isn't connected/connecting to an xDS server.
1 - Envoy is connecting/connected to the primary source.
2 - Envoy is connecting/connected to the failover source.

Risk Level: low.
Testing: Added integration tests.
Docs Changes: Will be added once the feature is complete
Release Notes: N/A.
Platform Specific Features: N/A

### CODE REVIEW

**Comment by adisuissa** [comment]
I think this should be a simple review.
/assign @nezdolik @htuch

File: source/extensions/config_subscription/grpc/grpc_stream.h
Comment (htuch) [inline source/extensions/config_subscription/grpc/grpc_stream.h:265]: Given there are only a few of these, would using the "text readout" stat type be clearer rather than magic numbers?

Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_stream.h:265]: We can add a text-readout for the connected source (I'd say that it should be the gRPC-address or Envoy Cluster name), but I'm not sure it will add much more given that per-xDS-type has a [specific readout](https://github.com/envoyproxy/envoy/blob/69ee8cf7b5f5d6843d2f20c0149d3f2e55ea9e66/source/extensions/config_subscription/grpc/grpc_mux_impl.cc#L330).
IMHO it is easier to build alerts on top of integral values rather than text, but LMK if you think otherwise.

Comment (htuch) [reply source/extensions/config_subscription/grpc/grpc_stream.h:265]: I guess integers are fine as long as they aren't too magical. Please add documentation for this new stat and its values, as they are now a permanent-ish contract.

**Comment by nezdolik** [comment]
otherwise lgtm

File: source/extensions/config_subscription/grpc/grpc_stream.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_stream.h:265]: Yeah, all the docs (including the stats docs) will be added in the next PR (after this and #35692 are merged).
The plan is to also flip the runtime flag to `true` after we run it internally for a few weeks and ensuring it all works well.

Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_stream.h:265]: FWIW, the docs describing the xDS-Failover feature are in #35746 (which is currently in draft mode until this is merged).

**Review by htuch — DISMISSED** [review]
Ready to merge when @nezdolik is good.

**Review by nezdolik — DISMISSED** [review]

File: source/extensions/config_subscription/grpc/grpc_stream.h
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_stream.h]: I've changed the type from an unsigned int to an enum.
I believe the code is more understandable this way (but maybe less readable with the explicit template types). LMK if you think I should revert this.

File: test/extensions/config_subscription/grpc/xds_failover_integration_test.cc
Comment (adisuissa) [reply test/extensions/config_subscription/grpc/xds_failover_integration_test.cc:313]: Yep, will be added in #35746

File: source/extensions/config_subscription/grpc/grpc_mux_impl.cc
Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_mux_impl.cc]: I've converted to an enum. Let me know if you think further explanation is needed.

File: source/extensions/config_subscription/grpc/grpc_stream.h
Comment (nezdolik) [inline source/extensions/config_subscription/grpc/grpc_stream.h:32]: can they simply be named `primary`, `failover`? :) not trying to block on this, just curious

Comment (adisuissa) [reply source/extensions/config_subscription/grpc/grpc_stream.h:32]: My thinking here is to reduce the abstraction layer penetration of the primary/failover to this class as much as possible (at least from code perspective, but added to the comments for maintainability sake). Generally speaking, a GrpcStream object doesn't need to know whether it is used with primary/failover support or not.
The alternative was to move the value setters to the classes that own the grpc stream, but that seemed to be a higher amount of refactoring (adding callbacks in a few places) that I'm not sure is justified as part of this work.

That said, I'm open to modifying the names if others think it is useful.

Comment (nezdolik) [reply source/extensions/config_subscription/grpc/grpc_stream.h:32]: Thanks for clarification.

**Review by nezdolik — APPROVED** [review]

**Review by htuch — APPROVED** [review]

**Pull request merged**
